sirepo/auth_db/user.py

```diff
@@ -34,40 +34,37 @@ def add_roles(self, roles, expiration=None):
         from sirepo import sim_data
 
         cls = self.__class__
-        with sirepo.util.THREAD_LOCK:
-            u = self.logged_in_user()
-            for r in roles:
-                try:
-                    self.new(uid=u, role=r, expiration=expiration).save()
-                except sqlalchemy.exc.IntegrityError:
-                    # role already exists
-                    pass
-            sim_data.audit_proprietary_lib_files(qcall=self.auth_db.qcall)
+        u = self.logged_in_user()
+        for r in roles:
+            try:
+                self.new(uid=u, role=r, expiration=expiration).save()
+            except sqlalchemy.exc.IntegrityError:
+                # role already exists
+                pass
+        sim_data.audit_proprietary_lib_files(qcall=self.auth_db.qcall)
 
     def add_role_or_update_expiration(self, role, expiration):
-        with sirepo.util.THREAD_LOCK:
-            if not self.has_role(role):
-                self.add_roles(roles=[role], expiration=expiration)
-                return
-            r = self.search_by(uid=self.logged_in_user(), role=role)
-            r.expiration = expiration
-            r.save()
+        if not self.has_role(role):
```

Comment:
auth_db executes an explicit [sqlalchemy.orm.Session.begin](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.begin):

> Begin a transaction, or nested transaction, on this [Session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session), if one is not already begun.

[7.0 Transaction Control At The SQL Level](https://www.sqlite.org/lockingv3.html) describes how SQLite handles this:

> The SQL command "BEGIN TRANSACTION" (the TRANSACTION keyword is optional) is used to take SQLite out of autocommit mode. Note that the BEGIN command does not acquire any locks on the database. After a BEGIN command, a SHARED lock will be acquired when the first SELECT statement is executed. A RESERVED lock will be acquired when the first INSERT, UPDATE, or DELETE statement is executed. No EXCLUSIVE lock is acquired until either the memory cache fills up and must be spilled to disk or until the transaction commits. In this way, the system delays blocking read access to the file until the last possible moment.

This explains the behavior that you were seeing a few days ago. The commit is what will cause the integrity error, and it doesn't happen except at the last possible moment. In normal databases, it would detect the integrity error in a transaction manager, but SQLite doesn't have any shared state except the file itself.

If two applications write simultaneously, one will get a constraint violation, possibly later than you would expect, which is why we should switch to a real database.